optimade/client/client.py

```diff
@@ -0,0 +1,564 @@
+                )
+            )
+            results = self._run(filter, endpoint, page_limit=None, paginate=True)
+            self.all_results[endpoint][filter] = results
+            return results
+
+    def count(
+        self, filter: str = None, endpoint: Optional[str] = None
+    ) -> Dict[str, int]:
+
+        if endpoint is None:
+            if self._current_endpoint is not None:
+                endpoint = self._current_endpoint
+                self._current_endpoint = None
+            else:
+                endpoint = "structures"
+
+        if filter is None:
+            filter = ""
+
+        with self._progress:
+            self._progress.print(
+                Panel(
+                    f"Counting results for filter [bold yellow]{endpoint}[/bold yellow]/[bold magenta][i]{filter}[/i][/bold magenta]",
+                    expand=False,
+                )
+            )
+            results = self._run(filter, endpoint, page_limit=1, paginate=False)
+            count_results = {}
+
+            for base_url in results:
+                count_results[base_url] = results[base_url].meta.get(
+                    "data_returned", None
+                )
+
+            self.count_results[endpoint][filter] = count_results
+            return count_results
+
+    def _run(self, filter, endpoint, page_limit, paginate) -> Dict[str, QueryResults]:
+
+        if self.use_async:
+            # Check for a preexisting event loop (e.g. within a Jupyter notebook)
+            # and use it if present
+            try:
+                event_loop = asyncio.get_running_loop()
+            except RuntimeError:
+                event_loop = None
+            if event_loop and event_loop.is_running():
+                task = asyncio.run_coroutine_threadsafe(
+                    self._get_all_async(
+                        endpoint, filter, page_limit=page_limit, paginate=paginate
+                    ),
+                    event_loop,
+                )
+                # This seems to hang on some Jupyter versions
+                results = task.result(timeout=10)
+            else:
+                results = asyncio.run(
+                    self._get_all_async(
+                        endpoint, filter, page_limit=page_limit, paginate=paginate
+                    )
+                )
+        else:
+            results = self._get_all(
+                endpoint, filter, page_limit=page_limit, paginate=paginate
+            )
+
+        return results
+
+    def get_one(
+        self,
+        endpoint: str,
+        filter: str,
+        base_url: str,
+        page_limit: Optional[int] = None,
+        paginate: bool = True,
+    ) -> Dict[str, QueryResults]:
+        try:
+            return self._get_one(
+                endpoint, filter, base_url, page_limit=page_limit, paginate=paginate
+            )
+        except (RuntimeError, httpx.TimeoutException, json.JSONDecodeError):
+            return {}
```

Comment:
Would it not be better to give some kind of warning that the request was unsuccessful and why it was unsuccessful?

Reply:
Yes, I'm going through and adding a few things like that now. The general API should remain constant though. I will probably make all functions except `get` and `count` private as they are not really meant to be called by the user.